asq-exercise-progress.html

```diff
@@ -0,0 +1,509 @@
+    * @param {number} Duration of the animation cycle, given in seconds.
+    */
+    _setVelocity:  function(property, velocity) {
+      var that = this;
+      this.customStyle[property] = velocity + 's';
+      this.updateStyles();
+      setTimeout(function() {
+        that.customStyle[property] = '0s';
+        that.updateStyles();
+      }, velocity * 1000);
+    },
+
+    /**
+    * Animates the progress bar every 5 seconds depending on the activity.
+    *
+    */
+    _visualizeChange: function() {
+      var that = this;
+
+      setInterval(function() {
+        var submissionChange = that._calcChange(that._submissionCounter);
+        var workingChange = that._calcChange(that._workingCounter);
+        var focusedChange = that._calcChange(that._focusedCounter);
+        var idleChange = that._calcChange(that._idleCounter);
+
+        that._setVelocity('--submitted-bar-duration', that._determineVelocity(submissionChange));
+        that._setVelocity('--working-bar-duration',  that._determineVelocity(workingChange));
+        that._setVelocity('--focused-bar-duration',  that._determineVelocity(focusedChange));
+        that._setVelocity('--idle-bar-duration',  that._determineVelocity(idleChange));
+
+        that._submissionCounter = 0;
+        that._workingCounter = 0;
+        that._focusedCounter = 0;
+        that._idleCounter = 0;
+      }, 5000);
+    },
+
+    /**
+    * Calculate the amount of change of a given activity.
+    *
+    * @param {number} A number that is smaller than or equal to the total number of users.
+    * @return {number} The amount of change in percent and in proportion to the number of users.
+    */
+    _calcChange: function(num) {
+      var total = this._submissionsNum + this._workingViewersNum + this._focusedViewersNum + this._idleViewersNum;
+      return num/total * 100;
+    },
+
+    /**
+    * Get the velocity for the paper progress indeterminate cycle.
+    *
+    * @param {number} A number in percent.
+    * @return {number} Returns a number in seconds for the duration of one paper progress
+    * indeterminate cycle.
+    */
+    _determineVelocity: function(num) {
+      if (num == 0) return 0;
+      else if(num < 5) return 10;
+      else if(num < 15) return 8;
+      else if(num < 30) return 6;
+      else if(num < 50) return 4;
+      else if(num < 75) return 2;
+      else if(num <= 100) return 1;
+      else return 0 // change was higher than 100%
+    },
+
+    created: function() {
+      document.addEventListener('asq-ready', function(evt) {
+        try {
+          this._subscribeToEvents(evt.detail.asqEventBus);
+        } catch (err) {
+          console.debug('failed to _subscribeToEvents');
+        }
+      }.bind(this));
+      this._visualizeChange();
+    },
+
+    _subscribeToEvents: function(eventBus) {
+      eventBus.on('asq:connected-clients', this._onConnectedClients.bind(this));
+      eventBus.on('asq:sessionEventBeamer', function(evt) {
```

Comment:
refactor the event handler to a named function like ` this._onConnectedClients.bind(this)`